File: typescript-kata/src/data-structures/linked-list.ts
class LNode<T> {
    val: T
    prev?: LNode<T>
    next?: LNode<T>

    constructor(val: T) {
        this.val = val
    }
}

export class LinkedList<T> {
    length: number
    head?: LNode<T>
    tail?: LNode<T>

    constructor() {
        this.head = this.tail = undefined
        this.length = 0
    }

    getLength(): number {
        return this.length
    }

    prepend(val: T): void {
        const node = new LNode(val)
        if (!this.head) {
            this.head = this.tail = node
            this.length++
            return
        }

        node.next = this.head;
        this.head.prev = node
        this.head = node
        this.length++

        return
    }

    append(val: T):void {
        const node = new LNode(val)
        if (!this.head) {
            this.head = this.tail = node
            this.length++
            return
        }

        node.prev = this.tail
        this.tail!.next = node
        this.tail = node
        this.length++

        return
    }

    get(indx: number): T | undefined {
        if (!this.head || indx >= this.getLength()) {
            return undefined
        }

        let curr: LNode<T> | undefined = this.head
        for (let i = 0; i < indx; ++i) {
            if (curr) {
                curr = curr.next
            }
        }

        return curr ? curr.val : undefined
    }

    private getNode(indx: number): LNode<T> | undefined {
        if (!this.head || indx >= this.getLength()) {
            return undefined
        }

        let curr: LNode<T> | undefined = this.head
        for (let i = 0; i < indx; ++i) {
            if (curr) {
                curr = curr.next
            }
        }

        return curr
    }

    insertAt(val: T, indx: number): void {
        if (indx > this.getLength()) {
            return
        };

        if (indx === 0) {
            this.prepend(val)
        }

        if (indx === this.getLength()) {
            this.append(val)
        }


        const oldNode = this.getNode(indx);
        if (!oldNode || !oldNode.prev || !oldNode.next) {
            return
        }

        const newNode = new LNode(val);

        newNode.next = oldNode
        newNode.prev = oldNode.prev
        oldNode.prev.next = newNode
        oldNode.prev = newNode
        this.length++
    }


    removeAt(indx: number): T | undefined {
        if (indx >= this.getLength() || indx < 0 || !this.head || !this.tail) {
            return
        }

        const nodeToRemove = this.getNode(indx)
        if (!nodeToRemove) {
            return
        }

        if (this.getLength() === 1) {
            this.head = this.tail = undefined
            this.length--
            return nodeToRemove.val
        }

        if (indx === 0) {
            this.head = this.head.next
            this.head!.prev = undefined
            this.length--
            return nodeToRemove.val
        }

        if (indx === this.getLength() - 1) {
            this.tail = this.tail.prev
            this.tail!.next = undefined
            this.length--
            return nodeToRemove.val
        }

        const prevNode = nodeToRemove.prev
        const nextNode = nodeToRemove.next

        if (!prevNode || !nextNode) {
            return
        }

        prevNode.next = nextNode
        nextNode.prev = prevNode


        this.length--

        // free memory
        nodeToRemove.next = undefined
        nodeToRemove.prev = undefined

        return nodeToRemove.val
    }

    remove(item: T): T | undefined {
        let nodeToRemove = this.head
        let indx = 0;
        while (nodeToRemove) {
            if (nodeToRemove.val === item) {
                break
            }
            nodeToRemove = nodeToRemove.next
            indx++
        }
        if (!nodeToRemove || !this.head || !this.tail) {
            return undefined
        }


        if (this.getLength() === 1) {
            this.head = this.tail = undefined
            this.length--
            return nodeToRemove.val
        }

        if (indx === 0) {
            this.head = this.head.next
            this.head!.prev = undefined
            this.length--
            return nodeToRemove.val
        }

        if (indx === this.getLength() - 1) {
            this.tail = this.tail.prev
            this.tail!.next = undefined
            this.length--
            return nodeToRemove.val
        }

        const prevNode = nodeToRemove.prev
        const nextNode = nodeToRemove.next

        if (!prevNode || !nextNode) {
            return
        }

        prevNode.next = nextNode
        nextNode.prev = prevNode


        this.length--

        // Free memory
        nodeToRemove.next = undefined
        nodeToRemove.prev = undefined

        return nodeToRemove.val
    }
}

File: python_kata/src/data_structures/linked_list.py
from typing import Generic, TypeVar

T = TypeVar("T")

class Node(Generic[T]):
    def __init__(self, val=None):
        self.val: T | None = val
        self.next: Node[T] | None = None
        self.prev: Node[T] | None = None 


class LinkedList(Generic[T]):
    def __init__(self):
        self.head: Node[T] | None = None 
        self.tail: Node[T] | None = None
        self.length: int = 0

    def get_length(self) -> int:
        return self.length

    def prepend(self, data) -> None:
        new_node = Node(data)

        if not self.head:
            self.head = self.tail = new_node
            self.length += 1
            return

        new_node.next = self.head
        self.head.prev = new_node

        self.head = new_node
        self.length += 1
        return

    def append(self, data) -> None:
        new_node = Node(data)
        if not self.head or not self.tail:
            self.head = self.tail = new_node
            self.length += 1
            return

        self.tail.next = new_node
        new_node.prev = self.tail
        self.tail = new_node

        self.length += 1
        return

    def get(self, indx: int) -> None | T:
        if indx >= self.get_length():
            return

        curr = self.head 
        for _ in range(0, indx, 1):
            if curr:
                curr = curr.next
        if not curr:
            return
        return curr.val

    def __getNode(self, indx: int) -> None | Node[T]:
        if indx >= self.get_length():
            return

        curr = self.head 
        for _ in range(0, indx, 1):
            if curr:
                curr = curr.next
        if not curr:
            return
        return curr

    def insert_at(self, data,  indx: int) -> None:
        if indx == 0:
            return self.prepend(data)

        if indx == self.get_length():
            return self.append(data)

        curr = self.__getNode(indx)

        if not curr or not curr.prev or not curr.next:
            return

        new_node = Node(data)

        new_node.next = curr
        new_node.prev = curr.prev

        curr.prev.next = new_node
        curr.prev = new_node

        self.length += 1

    def remove_at(self, indx: int) -> T | None :
        curr = self.__getNode(indx)
        if not curr:
            return None
        if self.get_length() == 1:
            self.head = None
            self.tail = None
            self.length = 0
            return curr.val 

        if indx == 0:
            if self.head != None and self.head.next != None:
                new_head = self.head.next
                new_head.prev = None
                self.head = new_head

                self.length -= 1
                return curr.val
        if indx == self.get_length() - 1:
            if self.tail != None and self.tail.prev != None:
                new_tail = self.tail.prev
                new_tail.next = None
                self.tail = new_tail

                self.length -= 1
                return


        if not curr or not curr.prev or not curr.next:
            return

        curr.prev.next = curr.next
        curr.next.prev = curr.prev

        # Free memory
        curr.next = None
        curr.prev = None

        return curr.val

    def remove(self, data):
        if not self.head:
            return

        curr = self.head
        while curr:
            if curr.val == data:
                break
            curr = curr.next

        if curr == None:
            return

        if self.length == 1:
            self.head = None
            self.tail = None
            self.length = 0

        if curr == self.head:
            if not curr.next:
                return
            curr.next.prev = None
            self.head = curr.next
            self.length -= 1
            return

        if curr == self.tail:
            if not curr.prev:
                return
            curr.prev.next = None
            self.tail = curr.prev
            self.length -= 1
            return


        if not curr.next or not curr.prev:
            return

        curr.prev.next = curr.next
        curr.next.prev = curr.prev

        # Free memory
        curr.next = None
        curr.prev = None

        return curr.val
